### PR TITLE
Add transformation from and to the `ghc-lib` AST data structure

### LINKED ISSUE #33: Implement `ghc-lib` as an alternative front end
## Description
Building on top of the new AST data structure implemented in #30, transformations from the `ghc-lib` AST data structure to our own structure and the other way around should be implemented, so that, in combination with corresponding command line option (#32), `ghc-lib` (or [`ghc-lib-parser`](https://hackage.haskell.org/package/ghc-lib-parser), to be more precise) can be used as an alternative front end.

## Motivation
The main motivation for the switch to a new AST data structure was the opportunity of using different parsers as alternative front ends. After implementing our own AST data structure and `haskell-src-exts` (the parser which we were already using previously) as one of these alternatives, the next step is to implement `ghc-lib`, which includes the parser used by the GHC itself, as the other alternative. Because of this connection to the GHC, `ghc-lib` is a more stable foundation for `haskell-src-transformations`, while the support for `haskell-src-exts` could drop at some point.

## Tasks
- [x] Take a look at the AST data structure of `ghc-lib` (a good starting point would be the type representing a Haskell module, which can be found [here](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/GHC-Hs.html#t:HsModule))
  + [x] Analyze how the types in both AST data structures correspond with each other
  + [x] If necessary, make changes to our own AST data structure (for example because of information missing in the `ghc-lib` structure)
- [x] Implement the transformation from the AST data structure of `ghc-lib` to our own
- [x] Implement the missing `Eq` and `Show` instances of the `ghc-lib` type variables (the [`ghc-lib-parser-ex`](https://hackage.haskell.org/package/ghc-lib-parser-ex) module could help a bit with that)
- [x] Implement the transformation from our own AST data structure to the one of `ghc-lib`
- [x] Add command line option to be able to use the `ghc-lib` front end (see #32 and #35)
- [x] Add the `ghc-lib` part of the `processInput` function in `Main.hs` (parsing and pretty printing the module)
- [x] Let the implemented tests run with the `ghc-lib` front end as well

### PR DESCRIPTION
**Pull request by stu204767**
### Issue
Closes #33.

### Description of the Change
Adds a two-way transformation between the AST data structure used by `ghc-lib` and our own AST data structure.

### Verification Process
The transformation has been tested in #43.

### Additional Notes
There are some constructs (for example statements in guards) where I have tested how they are represented in the `ghc-lib` AST for basic cases, but where I'm not sure if this is the only possible representation that we can support. I've marked these places with `TODO` comments in the code.

### CODE REVIEW

File: haskell-src-transformations.cabal [hunk 1]
```diff
@@ -31,14 +31,16 @@ flag PolysemyPluginEnabled {
...
-                     , haskell-src-exts >= 1.23  && < 1.24
-                     , mtl              >= 2.2   && < 2.3
-                     , polysemy         >= 1.3   && < 1.4
-                     , polysemy-plugin  >= 0.2.5 && < 0.2.6
+  build-depends:       base              >= 4.12      && < 4.13
+                     , containers        >= 0.6       && < 0.7
+                     , directory         >= 1.3       && < 1.4
+                     , extra             >= 1.7       && < 1.8
+                     , ghc-lib-parser    >= 8.6       && < 9.0
+                     , ghc-lib-parser-ex >= 8.10.0.15 && < 9.0
```
Comment (just95) [inline haskell-src-transformations.cabal]: The version constraint `8.10.0.15` is oddly specific. Doesn't `8.10` suffice? Is backward compatibility of all versions up to `9.0` guaranteed by `ghc-lib`'s versioning scheme? If not, the upper bounds should be `8.7` and `8.11`, respectively.

File: haskell-src-transformations.cabal [hunk 2]
```diff
@@ -84,6 +87,8 @@ library
                     , HST.Frontend.Syntax
                     , HST.Frontend.FromHSE
                     , HST.Frontend.ToHSE
+                    , HST.Frontend.FromGHC
+                    , HST.Frontend.ToGHC
```
Comment (just95) [inline haskell-src-transformations.cabal]:
```suggestion
                    , HST.Frontend.FromGHC
                    , HST.Frontend.FromHSE
                    , HST.Frontend.Syntax
                    , HST.Frontend.ToGHC
                    , HST.Frontend.ToHSE
```

File: src/lib/HST/Frontend/FromGHC.hs [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+
+-------------------------------------------------------------------------------
+-- Wrappers for @ghc-lib-parser@ Types                                       --
+-------------------------------------------------------------------------------
+
+-- | Wrapper for the two literal types (for regular and overloaded literals)
+--   used by @ghc-lib-parser@.
+data LitWrapper = Lit (GHC.HsLit GHC.GhcPs)
+                | OverLit (GHC.HsOverLit GHC.GhcPs)
+  deriving Eq
```
Comment (just95) [inline src/lib/HST/Frontend/FromGHC.hs]: Shouldn't literals also be compared with `defaultPrintEq` such that source spans are ignored?

File: src/lib/HST/Frontend/FromGHC.hs [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+    _ -> error "Expected a built-in special constructor"
+
+-- | Takes an HST source span and maps GHC names to HST special constructors.
+specialConMap :: S.SrcSpan GHC -> [(GHC.Name, S.SpecialCon GHC)]
+specialConMap s =
+  [ (GHC.tyConName GHC.unitTyCon          , S.UnitCon s)
+  , (GHC.listTyConName                    , S.ListCon s)
+  , (GHC.tyConName GHC.funTyCon           , S.FunCon s)
+  , (GHC.consDataConName                  , S.Cons s)
+  , (GHC.tupleTyConName GHC.UnboxedTuple 1, S.UnboxedSingleCon s)
```
Comment (just95) [inline src/lib/HST/Frontend/FromGHC.hs]: `S.UnboxedSingleCon` is `(# #)` but `GHC.tupleTyConName GHC.UnboxedTuple 1` sounds like `(# τ #)` to me.

File: src/lib/HST/Frontend/FromGHC.hs [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+  ]
```
Comment (just95) [inline src/lib/HST/Frontend/FromGHC.hs]: That the special case for `S.TupleCon` is not covered by `specialConMap` but is implemented in `transformName` should at least be documented. However, for error reporting purposes it would be better if all names were transformed in `transformName` using pattern matching.

File: src/lib/HST/Frontend/Syntax.hs [hunk 6]
```diff
@@ -78,9 +78,9 @@ deriving instance ShowAST a => Show (Decl a)
 
 -- | A data constructor. Does not include a source span and should not be
 --   transformed back.
-data ConDecl a = ConDecl (Name a) [TypeExp a]
-               | InfixConDecl (TypeExp a) (Name a) (TypeExp a)
-deriving instance EqAST a => Eq (ConDecl a)
+data ConDecl a = ConDecl (Name a) Int
```
Comment (stu204146-kiel) [inline src/lib/HST/Frontend/Syntax.hs]: Why did you remove the `TypeExp`s in `ConDecl`?

File: src/lib/HST/Frontend/FromGHC.hs [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+  GHC.showSDocOneLine defaultDynFlags (GHC.showAstData GHC.NoBlankSrcSpan d)
+
+-- | Returns default variants of the dynamic flags used by @ghc-lib-parser@.
+--
+--   These dynamic flags are used for parsing Haskell code and printing AST
+--   components.
+defaultDynFlags :: GHC.DynFlags
+-- TODO These DynFlags should be the same as the ones used for parsing Haskell
+-- modules and should be defined in another module.
+defaultDynFlags = GHC.defaultDynFlags GHC.fakeSettings GHC.fakeLlvmConfig
```
Comment (just95) [inline src/lib/HST/Frontend/FromGHC.hs]: Since the type family instances start to become more complicated, I'd suggest moving them to their own modules (e.g. `HST.Frontend.Syntax.GHC` and `HST.Frontend.Syntax.HSE`)

File: src/lib/HST/Frontend/Syntax.hs [hunk 8]
```diff
@@ -78,9 +78,9 @@ deriving instance ShowAST a => Show (Decl a)
...
+               | InfixConDecl (Name a)
```
Comment (just95) [inline src/lib/HST/Frontend/Syntax.hs]: Only the `ConDecl` constructor and a `isInfix` flag would probably suffice, too.

**Review by just95 — CHANGES_REQUESTED** [review]

File: haskell-src-transformations.cabal (hunk 1 above)
Comment (stu204767) [reply haskell-src-transformations.cabal]: The reason for this long version number is that `8.10` is the GHC version that is being built upon. `ghc-lib-parser-ex` is a fairly new package and has been restructured quite a bit under the GHC version `8.10`, which is why we have to be this specific. The upper bound is therefore very optimistic and should propably be `8.10.0.16` to be sure.

File: src/lib/HST/Frontend/FromGHC.hs (hunk 3 above)
Comment (stu204767) [reply src/lib/HST/Frontend/FromGHC.hs]: These literal types don't appear to contain source span information.

File: src/lib/HST/Frontend/FromGHC.hs (hunk 5 above)
Comment (stu204767) [reply src/lib/HST/Frontend/FromGHC.hs]: I wanted to use pattern matching, but the [`Name`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/Name.html#v:Name) and [`OccName`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/OccName.html#t:OccName) types of `ghc-lib-parser` don't export their constructors, which is why I chose this solution.
I will mention the special case for tuple constructors in the documentation, though.

File: src/lib/HST/Frontend/FromGHC.hs (hunk 4 above)
Comment (stu204767) [reply src/lib/HST/Frontend/FromGHC.hs]: Okay, I thought `SingleCon` meant one element, but I will change that to a `0` then.
All of these special constructors need to be tested once we start testing with `ghc-lib` (I wasn't able to do this yet, because most AST components of `ghc-lib-parser` don't have `Show` instances and the `showAstData` function simplifies the representation of names).

File: src/lib/HST/Frontend/Syntax.hs (hunk 8 above)
Comment (stu204767) [reply src/lib/HST/Frontend/Syntax.hs]: Do you want a new type for this flag or a `Bool`?
Either way, I don't think two constructors for `ConDecl` or a flag make much of a difference, but I will change it if you want.

File: src/lib/HST/Frontend/FromGHC.hs (hunk 7 above)
Comment (stu204767) [reply src/lib/HST/Frontend/FromGHC.hs]: I would call these modules `HST.Frontend.GHC` and `HST.Frontend.HSE`, so without `Syntax`, because they are not a part of our syntax.
We could also create subdirectories `GHC` and `HSE` in the `Frontend` directory and call these modules `HST.Frontend.GHC.GHCConfig` and `HST.Frontend.GHC.HSEConfig` or similar.

File: src/lib/HST/Frontend/Syntax.hs (hunk 8 above)
Comment (just95) [reply src/lib/HST/Frontend/Syntax.hs]: > I don't think two constructors for `ConDecl` or a flag make much of a difference

I agree. The only difference it would make is that we could reduce some code duplication in `HST.Application`.

> Do you want a new type for this flag or a Bool?

I'd suggest something like this.

```suggestion
data ConDecl a
  = ConDecl { conDeclName    :: Name a
            , conDeclArity   :: Int
            , conDeclIsInfix :: Bool
            }
```

File: src/lib/HST/Frontend/FromGHC.hs (hunk 7 above)
Comment (just95) [reply src/lib/HST/Frontend/FromGHC.hs]: I considered the `type instance`s part of the syntax since the `type family` declarations are in the `Syntax` module, too. However, `GHC` and `HSE` subdirectories are fine as well. `From*` and `To*` could be moved to the subdirectories as well. I'd just avoid using `GHC` and `HSE` in module names twice, i.e., `HST.Frontend.GHC.Config` and `HST.Frontend.GHC.From` instead of `HST.Frontend.GHC.GHCConfig` and `HST.Frontend.GHC.FromGHC`.

File: src/lib/HST/Frontend/FromGHC.hs [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+  in  (S.UnQual s' (S.Ident s' (GHC.occNameString name)), GHC.isDataOcc name)
+transformRdrName (GHC.L s (GHC.Qual modName name)) =
+  let s' = transformSrcSpan s
+  in  ( S.Qual s'
+               (transformModuleName s' modName)
+               (S.Ident s' (GHC.occNameString name))
+      , GHC.isDataOcc name
+      )
+transformRdrName (GHC.L s (GHC.Exact name)) =
+  let s' = transformSrcSpan s in (S.Special s' (transformName s' name), False)
```
Comment (just95) [inline src/lib/HST/Frontend/FromGHC.hs]: I was testing with the queue `./example`s and got the following error message whenever the module contained patterns for built-in data types.

> ```
> Internal: Only constructors can be applied in patterns
> CallStack (from HasCallStack):
>   error, called at src/lib/HST/Frontend/FromGHC.hs:324:14 in haskell-src-transformations-0.1.1.0-inplace:HST.Frontend.FromGHC
> ```

As far as I can tell, the problem is that this function returns `False` for special constructors. Do you think that it is safe to change this to `True`?

```suggestion
  let s' = transformSrcSpan s in (S.Special s' (transformName s' name), True)
```

File: src/lib/HST/Frontend/FromGHC.hs [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+-- | Transforms a GHC located unqualified reader name into an HST name.
+transformRdrNameUnqual :: GHC.Located GHC.RdrName -> S.Name GHC
+transformRdrNameUnqual (GHC.L s (GHC.Unqual occName)) =
+  S.Ident (transformSrcSpan s) (GHC.occNameString occName)
+transformRdrNameUnqual _ = error "Expected an unqualified name"
+
+-- | Transforms a GHC name with an HST source span into an HST special
+--   constructor.
+transformName :: S.SrcSpan GHC -> GHC.Name -> S.SpecialCon GHC
+transformName s name = case lookup name (specialConMap s) of
```
Comment (just95) [inline src/lib/HST/Frontend/FromGHC.hs]: With the patch above, unit and empty list patterns still caused the following error message in my tests.

> ```
> Internal: Expected a built-in special constructor
> CallStack (from HasCallStack):
>   error, called at src/lib/HST/Frontend/FromGHC.hs:376:10 in haskell-src-transformations-0.1.1.0-inplace:HST.Frontend.FromGHC
> ```

However, pair patterns worked correctly which suggests that the `lookup` in `specialConMap` probably does not work.

File: src/lib/HST/Frontend/FromGHC.hs (hunk 9 above)
Comment (stu204767) [reply src/lib/HST/Frontend/FromGHC.hs]: Yes, that should be safe. I didn't consider that these built-in types could appear in patterns as well.

File: src/lib/HST/Frontend/FromGHC.hs (hunk 10 above)
Comment (stu204767) [reply src/lib/HST/Frontend/FromGHC.hs]: Okay, then we should probably not use the `Eq` instance of [`Name`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/Name.html#t:Name) at all, it also has a note saying it should be used with caution.
What do you think about using `nameOccName :: Name -> OccName` and `occNameString :: OccName -> String` and implementing `specialConMap` on those strings?
Or, as this probably does retain the spelling of the built-in constructor, so `()` wouldn't be equal to `( )`, a pretty-printing function could be used instead.

File: src/lib/HST/Frontend/FromGHC.hs (hunk 9 above)
Comment (just95) [reply src/lib/HST/Frontend/FromGHC.hs]: See a69a8cf230c47c4c20293513a67c42e280f05ffd.

File: src/lib/HST/Frontend/FromGHC.hs (hunk 10 above)
Comment (just95) [reply src/lib/HST/Frontend/FromGHC.hs]: I did some more tests. As it turns out `ghc-lib` differentiates between names of type and data constructors. Unlike in `haskell-src-exts`, the list type constructor [`GHC.listTyConName`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/TysWiredIn.html#v:listTyConName) is not the same as the empty list constructor [`GHC.nilDataConName`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/TysWiredIn.html#v:nilDataConName), for example. Since we never need type constructors, I've changed `specialConMap` to use the data constructor names in 1af73e6c3df8cda6b987b2aa531f0254c244b52b. Furthermore, I had to change `GHC.ATyCon` to `GHC.AConLike` in `transformName` to support tuples in infix notation (e.g. `(,) x y`).

File: src/lib/HST/Frontend/GHC/From.hs [hunk 11]
```diff
@@ -0,0 +1,605 @@
...
+--   into the @haskell-src-transformations@ representation of a Haskell module.
+transformModule
+  :: Member Report r
+  => GHC.Located (GHC.HsModule GHC.GhcPs)
+  -> Sem r (S.Module GHC)
+transformModule (GHC.L s modul) =
+  let modName' = case GHC.hsmodName modul of
+        Just (GHC.L s' modName) ->
+          Just (transformModuleName (transformSrcSpan s') modName)
+        Nothing -> Nothing
```
Comment (just95) [inline src/lib/HST/Frontend/GHC/From.hs]: If you change the type of `transformModuleName` from `S.SrcSpan GHC -> GHC.ModuleName -> S.ModuleName GHC` to `GHC.Located GHC.ModuleName -> S.ModuleName GHC` you can simply write the following here.

```suggestion
  let modName' = transformModuleName <$> GHC.hsmodName modul
```

File: src/lib/HST/Frontend/GHC/To.hs [hunk 12]
```diff
@@ -0,0 +1,452 @@
...
+  return $ GHC.tyConName (GHC.tupleTyCon (transformBoxed boxed) arity)
+transformSpecialCon (S.Cons _) = return GHC.consDataConName
+transformSpecialCon (S.UnboxedSingleCon _) =
+  return $ GHC.tupleTyConName GHC.UnboxedTuple 0
+transformSpecialCon (S.ExprHole _) =
+  reportFatal
+    $  Message Error
+    $  "Encountered expression hole at name level in retransformation! "
+    ++ "(Expression holes should be transformed at expression level with "
+    ++ "the ghc-lib front end)"
```
Comment (just95) [inline src/lib/HST/Frontend/GHC/To.hs]:
```suggestion
  reportFatal
    $  Message Internal
    $  "Encountered expression hole at name level in retransformation. "
    ++ "Expression holes should be transformed at expression level with "
    ++ "the ghc-lib front end!"
```

File: src/lib/HST/Effect/Report.hs [hunk 13]
```diff
@@ -21,12 +21,17 @@ module HST.Effect.Report
...
   , reportToOutputOrCancel
   , reportToHandleOrCancel
   , filterReportedMessages
     -- * Interpretations for Other Effects
   , errorToReport
   , exceptionToReport
+  , failToReport
+    -- * Error Messages for AST Transformation
+  , notSupported
+  , skipNotSupported
```
Comment (just95) [inline src/lib/HST/Effect/Report.hs]: I don't think that it is a good idea to clutter this module with concrete error messages. Let's move them to their own module (e.g, `HST.Frontend.Common.Messages` or `HST.Frontend.Transformer.Messages`).

File: src/lib/HST/Frontend/GHC/From.hs [hunk 14]
```diff
@@ -0,0 +1,605 @@
...
+--   therefore skipped. @Nothing@ is returned if the data definition itself or
+--   any of its constructors is not supported.
+transformDataDefn
+  :: Member Report r
+  => GHC.HsDataDefn GHC.GhcPs
+  -> Sem r (Maybe [S.ConDecl GHC])
+transformDataDefn GHC.HsDataDefn { GHC.dd_cons = cons } = do
+  conDecls <- mapM transformConDecl cons
+  return $ if all isJust conDecls then Just (catMaybes conDecls) else Nothing
+transformDataDefn (GHC.XHsDataDefn _) = return Nothing
```
Comment (just95) [inline src/lib/HST/Frontend/GHC/From.hs]: From the [documentation of `XXHsDataDefn`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/src/GHC.Hs.Decls.html#HsDataDefn) it looks like the field of `XHsDataDefn` always has type `NoExtCon` (which has no constructors) if `GHC.GhcPs` is used. Thus, we could also write the following to statically ensure that we are not missing a constructor.

```suggestion
transformDataDefn (GHC.XHsDataDefn x) = noExtCon x
```

Otherwise, we should probably inform the user that the declaration is skipped. If no warning is printed, they might be surprised that constructors of some data types are not handled correctly.

```suggestion
transformDataDefn (GHC.XHsDataDefn _) = do
  skipNotSupported "Data type declaration extension"
  return Nothing
```

File: src/lib/HST/Frontend/GHC/From.hs [hunk 15]
```diff
@@ -0,0 +1,605 @@
...
+--   definitions are not supported by the pattern matching compiler and are
+--   therefore skipped. @Nothing@ is returned if the data definition itself or
+--   any of its constructors is not supported.
+transformDataDefn
+  :: Member Report r
+  => GHC.HsDataDefn GHC.GhcPs
+  -> Sem r (Maybe [S.ConDecl GHC])
+transformDataDefn GHC.HsDataDefn { GHC.dd_cons = cons } = do
+  conDecls <- mapM transformConDecl cons
+  return $ if all isJust conDecls then Just (catMaybes conDecls) else Nothing
```
Comment (just95) [inline src/lib/HST/Frontend/GHC/From.hs:188]: The [documentation of `HsDataDefn`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/src/GHC.Hs.Decls.html#HsDataDefn) suggests that the constructor can represent data family instances, too, which are not supported by us.

File: src/lib/HST/Frontend/GHC/From.hs [hunk 16]
```diff
@@ -0,0 +1,605 @@
...
+
+-- | Transforms a GHC located statement consisting only of a single expression
+--   into an HST expression.
+transformStmtExpr
+  :: Member Report r
+  => GHC.LStmt GHC.GhcPs (GHC.LHsExpr GHC.GhcPs)
+  -> Sem r (S.Exp GHC)
+transformStmtExpr (GHC.L _ (GHC.BodyStmt _ body _ _)) = transformExpr body
+transformStmtExpr _ = notSupported "Complex guards"
+-- TODO Are there more statements that can be safely converted to boolean expressions?
```
Comment (just95) [inline src/lib/HST/Frontend/GHC/From.hs]: I'd prefer to handle every constructor explicitly even if all of other constructors are `notSupported`.

File: src/lib/HST/Frontend/GHC/From.hs [hunk 17]
```diff
@@ -0,0 +1,605 @@
...
+transformExpr (GHC.L _ (GHC.HsBinTick _ _ _ _)) =
+  notSupported "Haskell program coverage"
+transformExpr (GHC.L _ (GHC.HsTickPragma _ _ _ _ _)) =
+  notSupported "Haskell program coverage"
+transformExpr (GHC.L _ (GHC.HsWrap _ _ _)) =
+  notSupported "Expressions introduced by the type checker"
+transformExpr (GHC.L _ (GHC.XExpr _)) = notSupported "Expression extensions"
+
+
+-- | Transforms a GHC located tuple argument consisting of an expression into
```
Comment (just95) [inline src/lib/HST/Frontend/GHC/From.hs]:
```suggestion

-- | Transforms a GHC located tuple argument consisting of an expression into
```

File: src/lib/HST/Frontend/GHC/From.hs [hunk 18]
```diff
@@ -0,0 +1,605 @@
...
+transformName s name = case lookup name (specialConMap s) of
+  Just sCon -> return sCon
+  Nothing   -> case GHC.wiredInNameTyThing_maybe name of
+    -- TODO Can the other TyThing constructors occur?
+    Just (GHC.ATyCon tc) -> case GHC.tyConTuple_maybe tc of
+      Just GHC.BoxedTuple -> return $ S.TupleCon s S.Boxed (GHC.tyConArity tc)
+      Just GHC.UnboxedTuple ->
+        return $ S.TupleCon s S.Unboxed (GHC.tyConArity tc)
+      _ -> notSupported "Unexpected special constructor"
+    _ -> notSupported "Expected a built-in special constructor"
```
Comment (just95) [inline src/lib/HST/Frontend/GHC/From.hs]: `notSupported` doesn't really fit in this case. However, since this code is changed in #43 anyway, we can probably leave it as is for now.

**Review by just95 — CHANGES_REQUESTED** [review]

File: src/lib/HST/Frontend/GHC/From.hs (hunk 15 above)
Comment (stu204767) [reply src/lib/HST/Frontend/GHC/From.hs:188]: Yes, but those should not occur under the `TyClD` constructor of [`HsDecl`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/GHC-Hs-Decls.html#t:HsDecl), but under `InstD` -> `DataFamInstD` -> `DataFamInstDecl`. This type also contains an `HsDataDefn`.

File: src/lib/HST/Frontend/GHC/To.hs (hunk 12 above)
Comment (stu204767) [reply src/lib/HST/Frontend/GHC/To.hs]: I suppose the other two `Message`s in this module should be `Internal` as well?

File: src/lib/HST/Frontend/GHC/From.hs (hunk 14 above)
Comment (stu204767) [reply src/lib/HST/Frontend/GHC/From.hs]: `noExtCon` can be used at almost all of these `X`-constructors, so I've changed this at all of those places. Interestingly, the extension of [`HsValBindsLR`](https://hackage.haskell.org/package/ghc-lib-parser-8.10.1.20200523/docs/GHC-Hs-Binds.html#t:HsValBindsLR) remains as the only one of these extensions that doesn't equal to `NoExtCon`.

File: src/lib/HST/Frontend/GHC/To.hs (hunk 12 above)
Comment (just95) [reply src/lib/HST/Frontend/GHC/To.hs]: Yes :+1:
An `Error` tells the user that they did something wrong whereas an `Internal` error points us to a bug.

File: src/lib/HST/Frontend/GHC/From.hs (hunk 18 above)
Comment (stu204767) [reply src/lib/HST/Frontend/GHC/From.hs]: That's the reason I haven't changed these messages.

**Review by just95 — APPROVED** [review]

**Review by stu204146-kiel — APPROVED** [review]

File: src/lib/HST/Frontend/Syntax.hs (hunk 6 above)
Comment (just95) [reply src/lib/HST/Frontend/Syntax.hs]: The `TypeExp`s are not needed since `ConDecl`s are not converted back to the original AST. The only information that we need about constructor declarations is their arity.
